dotcom-rendering/src/server/lib/logging-store.ts

```diff
@@ -0,0 +1,64 @@
+/**
+ * This module allows data about requests to be passed down to our logger without having to rely on prop drilling.
+ *
+ * This makes it easy for us to log info in a component deep in our component tree and have data such as the pageId and platform
+ * without the hassle of trying to get that data to it.
+ *
+ * We also take care of recording timings for data enhancement, and render time from this module and logging per request.
+ *
+ * For simplicity we're using Node 16's AsyncLocalStorage to handle state instead of any 3rd party dependency like Redux or React Context.
+ */
+
+import { AsyncLocalStorage } from 'async_hooks';
+
+type DCRLoggingStore = {
+	request: {
+		pageId: string;
+		path: string;
+		method: string;
+		type?: string;
+		platform?: string;
+	};
+	error?: {
+		message?: string;
+		stack?: string;
+	};
+};
+
+export const loggingStore = new AsyncLocalStorage<DCRLoggingStore>();
```

Comment:
Does this somehow understand the call stack that allows it to know exactly which request it is handling even without prop drilling?! I'd love to understand what's going on under the hood in node.js to make this happen!

Reply:
Yes! Assuming I'm understanding the docs correctly! It was a new class introduced in Node 16! Their example usecase is pretty much exactly what we're using it for.

https://nodejs.org/api/async_context.html#introduction

Afraid I'm not exactly sure how it works under the hood, nodejs dark magic!